File: app/src/main/java/com/example/personalaifoodmap/FoodMapActivity.java
package com.example.personalaifoodmap;

import static androidx.constraintlayout.helper.widget.MotionEffect.TAG;

import static java.lang.Double.parseDouble;
import static java.lang.Float.parseFloat;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.exifinterface.media.ExifInterface;
import androidx.fragment.app.FragmentManager;

import com.naver.maps.geometry.LatLng;
import com.naver.maps.map.LocationTrackingMode;
import com.naver.maps.map.MapFragment;
import com.naver.maps.map.NaverMap;
import com.naver.maps.map.OnMapReadyCallback;
import com.naver.maps.map.overlay.Marker;
import com.naver.maps.map.overlay.OverlayImage;
import com.naver.maps.map.util.FusedLocationSource;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class FoodMapActivity extends AppCompatActivity implements OnMapReadyCallback {

    private static final int PERMISSION_REQUEST_CODE = 100;
    private static final String[] PERMISSIONS = {
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION,
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.INTERNET
    };

    private FusedLocationSource mLocationSource;
    private NaverMap mNaverMap;
    ArrayList<PhotoData> pDataList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_foodmap);

        pDataList = (ArrayList<PhotoData>) getIntent().getSerializableExtra("uriArr");

        // 지도 객체 생성
        FragmentManager fm = getSupportFragmentManager();
        MapFragment mapFragment = (MapFragment)fm.findFragmentById(R.id.map);
        if (mapFragment == null) {
            mapFragment = MapFragment.newInstance();
            fm.beginTransaction().add(R.id.map, mapFragment).commit();
        }

        // getMapAsync를 호출하여 비동기로 onMapReady 콜백 메서드 호출
        // onMapReady에서 NaverMap 객체를 받음
        mapFragment.getMapAsync(this);

        // 위치를 반환하는 구현체인 FusedLocationSource 생성
        mLocationSource =
                new FusedLocationSource(this, PERMISSION_REQUEST_CODE);
    }

    @Override
    public void onMapReady(@NonNull NaverMap naverMap) {

        // 지도상에 마커 표시 (이미지의 위치 정보 기반)
        for (PhotoData photoData : pDataList) {
            float x;
            float y;
            x = photoData.lat;
            y = photoData.lon;

            // 초기값인 0이 아니면 마커 표시
            if (x != 0 && y != 0) {
                Marker marker = new Marker();
                marker.setPosition(new LatLng(x, y));

                marker.setMap(naverMap);
                marker.setWidth(100);
                marker.setHeight(100);
                marker.setIcon(OverlayImage.fromPath(photoData.uri));
                System.out.println(photoData.uri);
            }
        }

        // NaverMap 객체 받아서 NaverMap 객체에 위치 소스 지정
        mNaverMap = naverMap;
        mNaverMap.setLocationSource(mLocationSource);

        // 권한확인. 결과는 onRequestPermissionsResult 콜백 매서드 호출
        ActivityCompat.requestPermissions(this, PERMISSIONS, PERMISSION_REQUEST_CODE);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        // request code와 권한획득 여부 확인
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.length > 0
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                mNaverMap.setLocationTrackingMode(LocationTrackingMode.Follow);
            }
        }
    }

    @Override
    public void onBackPressed()
    {
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        return;

    }
}

File: app/src/main/java/com/example/personalaifoodmap/ui/activity/FoodMapActivity.kt
package com.example.personalaifoodmap.ui.activity

import android.Manifest
import android.content.ContentValues.TAG
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.viewModels
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDialog
import androidx.constraintlayout.widget.ConstraintLayout
import com.example.personalaifoodmap.viewmodels.FoodMapViewModel
import androidx.core.app.ActivityCompat
import androidx.core.net.toUri
import androidx.lifecycle.Observer
import com.example.personalaifoodmap.databinding.ActivityFoodMapBinding
import com.example.personalaifoodmap.FoodMapApplication
import com.example.personalaifoodmap.ui.MarkerRenderer
import com.example.personalaifoodmap.R
import com.example.personalaifoodmap.data.UserPhoto
import com.example.personalaifoodmap.viewmodels.FoodMapViewModelFactory
import com.google.android.gms.location.*
import com.google.android.gms.maps.*
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.maps.android.clustering.Cluster
import com.google.maps.android.clustering.ClusterManager
import com.google.maps.android.clustering.ClusterManager.OnClusterClickListener
import com.google.maps.android.clustering.ClusterManager.OnClusterItemClickListener
import com.google.maps.android.clustering.view.DefaultClusterRenderer
import java.util.ArrayList

internal class FoodMapActivity : AppCompatActivity(), OnMapReadyCallback{

    private lateinit var binding: ActivityFoodMapBinding
    private val foodMapViewModel : FoodMapViewModel by viewModels(){
        FoodMapViewModelFactory((application as FoodMapApplication).foodMapRepository)
    }

    private lateinit var mGoogleMap: GoogleMap
    private lateinit var currentLocation: Location
    private lateinit var clusterManager: ClusterManager<UserPhoto>
    lateinit var mapFragment:SupportMapFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityFoodMapBinding.inflate(layoutInflater)
        setContentView(binding.root)

        startFoodMap()
    }

    fun startFoodMap(){
        mapFragment = supportFragmentManager.findFragmentById(R.id.foodMapFragment) as SupportMapFragment
        mapFragment.getMapAsync(this)
    }

    override fun onMapReady(googleMap: GoogleMap) {

        // 지도 연결
        this.mGoogleMap = googleMap

        // 현재 위치 표시
        setCurrentLocation()

        // 클러스터링
        clusterManager = ClusterManager(this, mGoogleMap)
        mGoogleMap.setOnCameraIdleListener(clusterManager)
        clickCluster()
        clickClusterItem()

        // 마커 표시
        val markerRenderer = MarkerRenderer( this, mGoogleMap, clusterManager)
        clusterManager.renderer = markerRenderer

        foodMapViewModel.userFoodPhotos.observe(this, Observer {
            for(userPhoto in it){
                clusterManager.addItem(userPhoto)
            }
        })

    }

    fun clickCluster(){
        clusterManager.setOnClusterClickListener {
            val intent = Intent( this,ClusterListActivity::class.java )
            val clusterList = arrayListOf<UserPhoto>()
            for (photoData in it.items) {
                clusterList.add(photoData)
            }
            intent.putExtra("clusterList", clusterList)
            startActivity(intent)
            false
        }
    }

    fun clickClusterItem(){
        clusterManager.setOnClusterItemClickListener{
            val intent = Intent(this,RestaurantDetailActivity::class.java)
            intent.putExtra("resDetail", it)
            startActivity(intent)
            false
        }
    }

    fun setCurrentLocation(){
        checkLocationPermission()
        val locationRequest = LocationRequest().setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
        val locationCallback = object: LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult?) {
                locationResult?.let {
                    currentLocation = it.lastLocation
                    drawCurrentLocation()
                }
            }
        }

        val fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback, Looper.myLooper())
        mGoogleMap.isMyLocationEnabled = true
    }

    fun drawCurrentLocation(){
        val latLng = LatLng(currentLocation.latitude, currentLocation.longitude)
        mGoogleMap.moveCamera(CameraUpdateFactory.newLatLng(latLng))
        mGoogleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng,14f))
    }

    fun checkLocationPermission(){
        val permissionCode = 101
        if (ActivityCompat.checkSelfPermission(
                this, Manifest.permission.ACCESS_FINE_LOCATION) !=
            PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                this, Manifest.permission.ACCESS_COARSE_LOCATION) !=
            PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), permissionCode)
            return
        }
    }

}